ingestion/src/metadata/ingestion/source/database/database_source.py

```diff
@@ -0,0 +1,319 @@
+#  Copyright 2021 Collate
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  http://www.apache.org/licenses/LICENSE-2.0
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""
+Generic source to build database connectors.
+"""
+import traceback
+import uuid
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass
+from typing import Iterable, List, Optional, Tuple, Union
+
+from sqlalchemy.engine.reflection import Inspector
+
+from metadata.generated.schema.entity.data.database import Database
+from metadata.generated.schema.entity.data.databaseSchema import DatabaseSchema
+from metadata.generated.schema.entity.data.table import (
+    Column,
+    DataModel,
+    Table,
+    TableData,
+    TableProfile,
+)
+from metadata.generated.schema.type.entityReference import EntityReference
+from metadata.ingestion.api.common import Entity
+from metadata.ingestion.api.source import Source
+from metadata.ingestion.models.ometa_table_db import OMetaDatabaseAndTable
+from metadata.ingestion.models.ometa_tag_category import OMetaTagAndCategory
+from metadata.ingestion.models.table_metadata import DeleteTable
+from metadata.ingestion.ometa.utils import ometa_logger
+from metadata.orm_profiler.orm.converter import ometa_to_orm
+from metadata.orm_profiler.profiler.default import DefaultProfiler
+from metadata.utils.filters import filter_by_schema, filter_by_table
+from metadata.utils.fqdn_generator import get_fqdn
+
+logger = ometa_logger()
+
+
+@dataclass  # type: ignore[misc]
+class DatabaseSourceService(Source, metaclass=ABCMeta):
+    @abstractmethod
+    def get_databases(self) -> Iterable[Inspector]:
+        """
+        Method Yields Inspector objects for each aviable database
+        """
+        pass
+
+    @abstractmethod
+    def get_schemas(self, inspector: Inspector) -> str:
+        """
+        Method Yields schemas aviable in database
+        """
+        pass
+
+    @abstractmethod
+    def standardize_schema_table_names(
+        self, schema: str, table: str
+    ) -> Tuple[str, str]:
+        """
+        Method formats Table & Schema names if required
+        """
+        pass
+
+    @abstractmethod
+    def _get_table_description(
```

Comment:
I don't think we are using "private" methods consistently. Why do we have `get_schemas` but `_get_table_description`? Feels like we could remove the leading `_` in most of these